README.md

```diff
@@ -3,58 +3,153 @@ xqueue_watcher
+- **Individual Submission Grader**: each exercise or homework may specify its own "grader". This should map to a file on the server that usually specifies test cases or additional processing for the student submission.
+
+Usually your server will look like this:
+```
+root/
+├── xqueue-watcher/
+│   ├── ... # xqueue-watcher repo, unchanged
+│   └── ...
+├── config/
+│   └── conf.d/
+│   │   └── my-course.json
+│   └── logging.json
+└── my-course/
+   ├── exercise1/
+   │   ├── grader.py  # - per-exercise grader
+   │   └── answer.py  # - if using JailedGrader
+   ├── ...
+   └── exercise2/
+       ├── grader.py
+       └── answer.py
+```
+Running XQueue Watcher:
+======================
+
+Usually you can run XQueue Watcher without making any changes. You should keep course-specific in another folder like shown above, so that you can update xqueue_watcher anytime.
+
+Install the requirements before running `xqueue_watcher`
+```bash
+cd xqueue-watcher/
+make requirements
+```
+
+Now you're ready to run it.
+```bash
+python -m xqueue_watcher -d [path to the config directory, eg ../config]
+```
+
+The course configuration JSON file in `conf.d` should have the following structure:
+```json
+    {
+        "test-123": {
+            "SERVER": "http://127.0.0.1:18040",
+            "CONNECTIONS": 1,
+            "AUTH": ["lms", "lms"],
+            "HANDLERS": [
+                {
+                    "HANDLER": "xqueue_watcher.grader.Grader",
+                    "KWARGS": {
+                        "grader_root": "/path/to/course/graders/",
+                    }
+                }
+            ]
+        }
+    }
+```
 
 * `test-123`: the name of the queue
 * `SERVER`: XQueue server address
 * `AUTH`: list of username, password
 * `CONNECTIONS`: how many threads to spawn to watch the queue
 * `HANDLERS`: list of callables that will be called for each queue submission
-	* `HANDLER`: callable name
-	* `KWARGS`: optional keyword arguments to apply during instantiation
+   * `HANDLER`: callable name, see below for Submissions Handler
+   * `KWARGS`: optional keyword arguments to apply during instantiation
+      * `grader_root`: path to the course directory, eg /path/to/my-course
 
+> TODO: document logging.json
 
-xqueue_watcher.grader.Grader
-========================
+Submissions Handler
+===================
+
+When xqueue_watcher detects any new submission, it will be passed to the submission handler for grading. It will instantiate a new handler based on the name configured above, with submission information retrieved
+from XQueue. There is a base grader defined in xqueue_watcher: Grader and JailedGrader (for Python, using CodeJail). If you don't use JailedGrader, you'd have to implement your own Grader by subclassing `xqueue_watcher.grader.Grader
+
+The payload from XQueue will be a JSON that usually looks like this, notice that "grader" is a required field in the "grader_payload" and must be configured accordingly in the Studio for the exercise.
```

Comment:
Could you briefly explain how to configure in Studio?

Reply:
Nit: Maybe instead:
`The payload received from XQueue will be a JSON object that usually looks like the JSON below. Note that "grader" is a required field in the "grader_payload" and must be configured accordingly in Studio.`?